File: algorithms/2_medium/C/04044_count_good_cyclic_rotations/solution.c
int countGoodRotations(const int *restrict nums, const int len) {
    if (len <= 0 || (len & 1) == 1 || nums == NULL) { return -1; }

    const int half = len / 2;

    int64_t full_sum = 0;
    for (int i = 0; i < len; i += 1) { full_sum += nums[i]; }

    int64_t window_sum = 0;
    for (int i = half; i < len; i += 1) { window_sum += nums[i]; }

    int cnt = (2 * window_sum > full_sum ? 1 : 0);

    for (int i = 0; i < len - 1; i += 1) {
        window_sum -= nums[(i + len - half) % len];
        window_sum += nums[i];
        if (2 * window_sum > full_sum) {
            cnt += 1;
        }
    }

    return cnt;
}

File: algorithms/2_medium/C/04044_count_good_cyclic_rotations/solution.rs
impl Solution {
    pub fn count_good_rotations(nums: Vec<i32>) -> i32 {
        assert!((nums.len() & 1) == 0 && nums.len() > 0);
        let len = nums.len();
        let half = len / 2;
        let full_sum = nums.iter().map(|&x| x as i64).sum::<i64>();
        let mut window_sum = nums[half..len].iter().map(|&x| x as i64).sum::<i64>();
        let mut cnt = if 2 * window_sum > full_sum { 1 } else { 0 };

        for i in 0..(len - 1) {
            window_sum -= nums[(i + len - half) % len] as i64;
            window_sum += nums[i] as i64;
            if 2 * window_sum > full_sum { 
                cnt += 1;
            }
        }

        cnt
    }
}
